### ISSUE TITLE
Problems with autodoc and Python functools.partial and partialmethod

### ISSUE THREAD

**Issue by Hvass-Labs** [sphinx-doc/sphinx#7023]
Thanks for making sphinx! I am using sphinx and readthedocs to automatically generate docs whenever I push something to github. I am very impressed with the whole system!

I have a few minor problems relating to sphinx autodoc and Python's `functools.partial` and `functools.partialmethod`. It is easiest if I show you the examples:

### Problem 1:

Sphinx autodoc works fine with a single level of `functools.partial`, but when using it in a nested / recursive fashion, the docs are not shown in sphinx. See for example `load_fundamental` which uses `functools.partial` once, and then see `load_income` which uses `functools.partial` again on `load_fundamental`:

https://github.com/simfin/simfin/blob/master/simfin/load.py#L174-L182

But only the docs for `load_fundamental` are generated by sphinx:

https://simfin.readthedocs.io/en/latest/load.html#simfin.load.load_fundamental

### Problem 2:

Sphinx autodoc does not seem to work at all for `functools.partialmethod` for class methods. See for example `load_fundamental` and `load_income` in this class:

https://github.com/simfin/simfin/blob/master/simfin/hubs.py#L238-L264

Here `load_fundamental` is a regular class method which has sphinx docs generated, but `load_income` which is a `partialmethod` does not have any docs generated by sphinx:

https://simfin.readthedocs.io/en/latest/hubs.html#simfin.hubs.StockHub.load_fundamental

### Problem 3:

The functions created with `functools.partial` in the module `load.py` are imported in another Python module `hubs.py`, but sphinx also shows them as belonging to `hubs.py` when using this rst-file:

https://github.com/SimFin/simfin/blob/master/docs/hubs.rst

These are basically the contents of the rst-file:

    .. automodule:: simfin.hubs
       :members:

These two python modules are located here:

https://github.com/SimFin/simfin/blob/master/simfin/load.py
https://github.com/SimFin/simfin/blob/master/simfin/hubs.py

### Problem 4:

This problem is not related to `functools.partial` but it's a very small problem, so rather than opening a separate issue, I include it here.

There is a minor inconsistency in how sphinx indents the doc-strings for parameters. Consider this page again:

https://simfin.readthedocs.io/en/latest/load.html

For the function `load()` the doc-string for the first parameter `dataset` starts at a new line. This is also the case for the parameters `variant` and `market`. But for the parameter `parse_dates` the doc-string does not start with this newline.

If you look again at the Python code, it appears that sphinx inserts a newline at the beginning of the parameter's doc-string, IF there is a blank line somewhere in the doc-string for that parameter.

https://github.com/SimFin/simfin/blob/master/simfin/load.py

I hope you can fix these minor issues. Thanks again for making sphinx, it's a very cool system!

**Comment by tk0miya** [comment]
Thank you for reporting.

Problem 1 & 3:
Confirmed. Indeed they are bugs of autodoc. I'll fix them later. 

Problem 2:
This is not a bug. Sphinx considers a partialmethod `load_income` does not have docstring. So you need to give `:undoc-members:` option to autodoc or set docstring to it. As an easy solution, you can set docstring via `#:` style comment:
```
    #: Load Income Statements.
    #: Sphinx considers comments before the assignment as a docstring of the member.
    #: The comment should be starts with `#:` (not single `#`).
    load_income = partialmethod(load_fundamental, dataset='income')
```

Problem 4:
It was already reported as #2768. So I'll not work it in this issue.

**Comment by Hvass-Labs** [comment]
Thanks very much for the quick response. I see you have already fixed problem 1. That is greatly appreciated!

**Comment by tk0miya** [comment]
I merged fixes #7031 and #7023 for Problem 1 & 3.
Thank you for reporting!
